File: packages/node/src/cache/memory-key-cache.ts
import { type Clock, systemClock } from "./clock.js";

/** A loader resolves the fresh value for a cache key (e.g. fetch a JWKS). */
export type CacheLoader<V> = (key: string) => Promise<V>;

/** Options for {@link MemoryKeyCache}. */
export interface MemoryKeyCacheOptions<V> {
  /** How long a loaded value stays fresh, in milliseconds. */
  ttlMs: number;
  /** Resolves a fresh value for a key. Invoked at most once per key concurrently. */
  loader: CacheLoader<V>;
  /**
   * If a refresh of an already-cached key fails, keep serving the previous value
   * for up to this many ms past its TTL instead of throwing. Protects against
   * transient JWKS-endpoint blips. Set to `0` to disable. Defaults to `ttlMs`.
   */
  staleIfErrorMs?: number;
  /**
   * When set, a background timer proactively refreshes every live key on this
   * interval so request-path lookups rarely block on the network. The timer is
   * `unref`'d and never keeps the process alive. Call {@link MemoryKeyCache.dispose}
   * to stop it.
   */
  refreshIntervalMs?: number;
  /** Injectable clock for deterministic tests. Defaults to the system clock. */
  clock?: Clock;
  /** Optional hook for observability (e.g. logging a failed background refresh). */
  onError?: (key: string, error: unknown) => void;
}

interface CacheEntry<V> {
  value: V;
  /** Epoch ms after which the value is considered stale and must be refreshed. */
  expiresAt: number;
}

/**
 * An in-memory, per-key TTL cache with **single-flight** load deduplication.
 *
 * Guarantees:
 *  - **Single-flight:** N concurrent `get(key)` calls during a miss trigger the
 *    loader exactly once; all callers await the same in-flight promise.
 *  - **TTL + lazy eviction:** stale entries are reloaded on access; a periodic
 *    sweep and optional background refresh keep memory bounded and values warm.
 *  - **Stale-if-error:** a failed refresh of a previously-good key serves the old
 *    value within a grace window rather than failing open or hard-erroring.
 *
 * It is value-generic so it can cache JWKS key sets, OIDC metadata, etc. The core
 * has no knowledge of what `V` is — keeping this layer provider-agnostic.
 */
export class MemoryKeyCache<V> {
  private readonly entries = new Map<string, CacheEntry<V>>();
  private readonly inFlight = new Map<string, Promise<V>>();
  private readonly ttlMs: number;
  private readonly staleIfErrorMs: number;
  private readonly loader: CacheLoader<V>;
  private readonly clock: Clock;
  private readonly onError: ((key: string, error: unknown) => void) | undefined;
  private refreshTimer: ReturnType<typeof setInterval> | undefined;
  private disposed = false;

  public constructor(options: MemoryKeyCacheOptions<V>) {
    if (options.ttlMs <= 0) {
      throw new RangeError("MemoryKeyCache: ttlMs must be greater than 0");
    }
    this.ttlMs = options.ttlMs;
    this.staleIfErrorMs = options.staleIfErrorMs ?? options.ttlMs;
    this.loader = options.loader;
    this.clock = options.clock ?? systemClock;
    this.onError = options.onError;

    if (options.refreshIntervalMs !== undefined && options.refreshIntervalMs > 0) {
      this.refreshTimer = setInterval(() => {
        void this.refreshAll();
      }, options.refreshIntervalMs);
      // Do not keep the event loop alive solely for cache refreshes.
      this.refreshTimer.unref?.();
    }
  }

  /**
   * Resolve the value for `key`, loading it if absent or stale. Concurrent calls
   * for the same key share a single load (single-flight).
   */
  public async get(key: string): Promise<V> {
    this.assertNotDisposed();
    const entry = this.entries.get(key);
    if (entry !== undefined && !this.isExpired(entry)) {
      return entry.value;
    }
    return this.load(key);
  }

  /**
   * Force a reload of `key`, bypassing the freshness check but still
   * deduplicating with any in-flight load. Returns the fresh value.
   */
  public async refresh(key: string): Promise<V> {
    this.assertNotDisposed();
    return this.load(key, { force: true });
  }

  /** Return the cached value without loading. `undefined` if absent or expired. */
  public peek(key: string): V | undefined {
    const entry = this.entries.get(key);
    if (entry === undefined || this.isExpired(entry)) {
      return undefined;
    }
    return entry.value;
  }

  /** Whether a fresh (non-expired) entry exists for `key`. */
  public has(key: string): boolean {
    return this.peek(key) !== undefined;
  }

  /** Remove a single key's cached value. */
  public delete(key: string): void {
    this.entries.delete(key);
  }

  /** Remove every cached value. In-flight loads are left to settle. */
  public clear(): void {
    this.entries.clear();
  }

  /** Number of entries currently held (including expired-but-not-swept). */
  public size(): number {
    return this.entries.size;
  }

  /** Sweep and drop every expired entry. Returns the count removed. */
  public evictExpired(): number {
    let removed = 0;
    for (const [key, entry] of this.entries) {
      if (this.isExpired(entry)) {
        this.entries.delete(key);
        removed += 1;
      }
    }
    return removed;
  }

  /** Stop the background refresh timer and drop all state. Idempotent. */
  public dispose(): void {
    this.disposed = true;
    if (this.refreshTimer !== undefined) {
      clearInterval(this.refreshTimer);
      this.refreshTimer = undefined;
    }
    this.entries.clear();
    this.inFlight.clear();
  }

  // --- internals -----------------------------------------------------------

  private isExpired(entry: CacheEntry<V>): boolean {
    return this.clock.now() >= entry.expiresAt;
  }

  /**
   * The single-flight core. If a load for `key` is already running, return that
   * promise; otherwise start one, register it, and clean up on settle.
   */
  private load(key: string, opts: { force?: boolean } = {}): Promise<V> {
    const existing = this.inFlight.get(key);
    if (existing !== undefined) {
      return existing;
    }

    // Re-check freshness for non-forced loads: a value may have been populated
    // between the caller's `get` check and acquiring the single-flight slot.
    if (opts.force !== true) {
      const entry = this.entries.get(key);
      if (entry !== undefined && !this.isExpired(entry)) {
        return Promise.resolve(entry.value);
      }
    }

    const promise = this.loader(key)
      .then((value) => {
        this.entries.set(key, { value, expiresAt: this.clock.now() + this.ttlMs });
        return value;
      })
      .catch((error: unknown) => {
        // Stale-if-error: if we still hold a value within the grace window, serve
        // it rather than propagating a transient loader failure.
        const stale = this.entries.get(key);
        if (stale !== undefined && this.clock.now() < stale.expiresAt + this.staleIfErrorMs) {
          this.onError?.(key, error);
          return stale.value;
        }
        throw error;
      })
      .finally(() => {
        this.inFlight.delete(key);
      });

    this.inFlight.set(key, promise);
    return promise;
  }

  /** Background-refresh every currently-tracked key, swallowing per-key errors. */
  private async refreshAll(): Promise<void> {
    const keys = [...this.entries.keys()];
    await Promise.all(
      keys.map((key) =>
        this.refresh(key).catch((error: unknown) => {
          this.onError?.(key, error);
        }),
      ),
    );
  }

  private assertNotDisposed(): void {
    if (this.disposed) {
      throw new Error("MemoryKeyCache: cache has been disposed");
    }
  }
}

File: packages/go/cache/memory_key_cache.go
package cache

import (
	"errors"
	"sync"
	"time"
)

// ErrDisposed is returned by a cache method called after Dispose.
var ErrDisposed = errors.New("cache: cache has been disposed")

// Loader resolves the fresh value for a cache key (e.g. fetch a JWKS).
type Loader[V any] func(key string) (V, error)

// Options configures a MemoryKeyCache.
type Options[V any] struct {
	// TTL is how long a loaded value stays fresh. Required (> 0).
	TTL time.Duration

	// Loader resolves a fresh value for a key. Invoked at most once per key
	// concurrently (single-flight).
	Loader Loader[V]

	// StaleIfError: if a refresh of an already-cached key fails, keep serving the
	// previous value for up to this long past its TTL instead of erroring. Zero
	// defaults to TTL; a negative value disables it.
	StaleIfError time.Duration

	// RefreshInterval, when > 0, starts a background goroutine that proactively
	// refreshes every live key on this interval. Stop it with Dispose.
	RefreshInterval time.Duration

	// Clock is an injectable clock. Defaults to SystemClock.
	Clock Clock

	// OnError is an optional hook for observability (e.g. a failed background or
	// stale-served refresh).
	OnError func(key string, err error)
}

type entry[V any] struct {
	value     V
	expiresAt time.Time
}

type call[V any] struct {
	wg    sync.WaitGroup
	value V
	err   error
}

// MemoryKeyCache is an in-memory, per-key TTL cache with single-flight load
// deduplication, stale-if-error fallback, and optional background refresh.
//
// Guarantees:
//   - Single-flight: N concurrent Get(key) calls during a miss trigger the
//     loader exactly once; all callers receive the same result.
//   - TTL + lazy eviction: stale entries are reloaded on access.
//   - Stale-if-error: a failed refresh of a previously-good key serves the old
//     value within a grace window rather than failing.
type MemoryKeyCache[V any] struct {
	ttl          time.Duration
	staleIfError time.Duration
	loader       Loader[V]
	clock        Clock
	onError      func(string, error)

	mu       sync.Mutex
	entries  map[string]entry[V]
	inflight map[string]*call[V]
	disposed bool

	stop     chan struct{}
	stopOnce sync.Once
}

// New constructs a MemoryKeyCache. It panics if TTL <= 0.
func New[V any](opts Options[V]) *MemoryKeyCache[V] {
	if opts.TTL <= 0 {
		panic("cache: TTL must be greater than 0")
	}
	stale := opts.StaleIfError
	if stale == 0 {
		stale = opts.TTL
	}
	if stale < 0 {
		stale = 0
	}
	clk := opts.Clock
	if clk == nil {
		clk = SystemClock
	}
	c := &MemoryKeyCache[V]{
		ttl:          opts.TTL,
		staleIfError: stale,
		loader:       opts.Loader,
		clock:        clk,
		onError:      opts.OnError,
		entries:      make(map[string]entry[V]),
		inflight:     make(map[string]*call[V]),
		stop:         make(chan struct{}),
	}
	if opts.RefreshInterval > 0 {
		go c.refreshLoop(opts.RefreshInterval)
	}
	return c
}

// Get resolves the value for key, loading it if absent or stale. Concurrent
// calls for the same key share a single load.
func (c *MemoryKeyCache[V]) Get(key string) (V, error) {
	c.mu.Lock()
	if e, ok := c.entries[key]; ok && !c.isExpired(e) {
		v := e.value
		c.mu.Unlock()
		return v, nil
	}
	c.mu.Unlock()
	return c.load(key, false)
}

// Refresh forces a reload of key, bypassing the freshness check but still
// deduplicating with any in-flight load.
func (c *MemoryKeyCache[V]) Refresh(key string) (V, error) {
	return c.load(key, true)
}

// Peek returns the cached value without loading. ok is false if absent/expired.
func (c *MemoryKeyCache[V]) Peek(key string) (V, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if e, ok := c.entries[key]; ok && !c.isExpired(e) {
		return e.value, true
	}
	var zero V
	return zero, false
}

// Has reports whether a fresh (non-expired) entry exists for key.
func (c *MemoryKeyCache[V]) Has(key string) bool {
	_, ok := c.Peek(key)
	return ok
}

// Delete removes a single key's cached value.
func (c *MemoryKeyCache[V]) Delete(key string) {
	c.mu.Lock()
	delete(c.entries, key)
	c.mu.Unlock()
}

// Size returns the number of entries currently held (including expired-but-not-swept).
func (c *MemoryKeyCache[V]) Size() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.entries)
}

// EvictExpired drops every expired entry and returns the count removed.
func (c *MemoryKeyCache[V]) EvictExpired() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	removed := 0
	for k, e := range c.entries {
		if c.isExpired(e) {
			delete(c.entries, k)
			removed++
		}
	}
	return removed
}

// Dispose stops the background refresh timer and drops all state. Idempotent.
func (c *MemoryKeyCache[V]) Dispose() {
	c.mu.Lock()
	c.disposed = true
	c.entries = make(map[string]entry[V])
	c.mu.Unlock()
	c.stopOnce.Do(func() { close(c.stop) })
}

func (c *MemoryKeyCache[V]) isExpired(e entry[V]) bool {
	return !c.clock.Now().Before(e.expiresAt)
}

// load is the single-flight core.
func (c *MemoryKeyCache[V]) load(key string, force bool) (V, error) {
	c.mu.Lock()
	if c.disposed {
		c.mu.Unlock()
		var zero V
		return zero, ErrDisposed
	}
	// Re-check freshness for non-forced loads: a value may have been populated
	// between the caller's Get check and acquiring the single-flight slot.
	if !force {
		if e, ok := c.entries[key]; ok && !c.isExpired(e) {
			v := e.value
			c.mu.Unlock()
			return v, nil
		}
	}
	if cl, ok := c.inflight[key]; ok {
		c.mu.Unlock()
		cl.wg.Wait()
		return cl.value, cl.err
	}
	cl := &call[V]{}
	cl.wg.Add(1)
	c.inflight[key] = cl
	c.mu.Unlock()

	value, err := c.loader(key)

	c.mu.Lock()
	var onErr func(string, error)
	if err != nil {
		// Stale-if-error: if we still hold a value within the grace window, serve
		// it rather than propagating a transient loader failure.
		if e, ok := c.entries[key]; ok && c.clock.Now().Before(e.expiresAt.Add(c.staleIfError)) {
			cl.value, cl.err = e.value, nil
			onErr = c.onError
		} else {
			cl.err = err
		}
	} else {
		c.entries[key] = entry[V]{value: value, expiresAt: c.clock.Now().Add(c.ttl)}
		cl.value = value
	}
	delete(c.inflight, key)
	c.mu.Unlock()
	cl.wg.Done()

	if onErr != nil {
		onErr(key, err)
	}
	return cl.value, cl.err
}

func (c *MemoryKeyCache[V]) refreshLoop(interval time.Duration) {
	t := time.NewTicker(interval)
	defer t.Stop()
	for {
		select {
		case <-c.stop:
			return
		case <-t.C:
			c.mu.Lock()
			keys := make([]string, 0, len(c.entries))
			for k := range c.entries {
				keys = append(keys, k)
			}
			c.mu.Unlock()
			for _, k := range keys {
				if _, err := c.Refresh(k); err != nil && c.onError != nil {
					c.onError(k, err)
				}
			}
		}
	}
}
